File: 섬 연결하기.py
def find(n, parent):
    while parent[n] != n:
        n = parent[n]
        find(n,parent)
    return parent[n]

def union(a,b,parent):
    a = find(a, parent)
    b = find(b, parent)
    if a<b:
        parent[b]=a
    else:
        parent[a] = b
def solution(n, costs):
    answer = 0
    graph = []
    parent = [0 for _ in range(n)]
    for i in range(n):
        parent[i] = i 
    for i in costs:
        a,b,c = i
        graph.append((c,a,b))
    graph.sort()
    
    for i in graph:
        cost, a, b = i
        if find(a, parent) != find(b, parent):
            union(a,b,parent)
            answer += cost
    return answer

File: 섬 연결하기.java
import java.util.*;
class Solution {
    static public int find(int x, int [] parent){
        if (x != parent[x]){
            parent[x] = find(parent[x], parent);
        }
        return parent[x];
    }
    static public int[] union(int a, int b, int [] parent){
        a = find(a, parent);
        b = find(b, parent);
        if (b>a) parent[b] = a;
        else parent[a] = b;
        return parent;
    }
    public int solution(int n, int[][] costs) {
        int answer = 0;
        List<ArrayList<Integer>> q= new ArrayList<>();
        Arrays.sort(costs, (a,b)-> Integer.compare(a[2],b[2]));
        for (int [] i : costs){
            ArrayList<Integer> temp = new ArrayList<>();
            temp.add(i[2]);
            temp.add(i[0]);
            temp.add(i[1]);
            q.add(temp);
        }
        int [] parent = new int[n];
        for(int i=0; i<n;i++){
            parent[i] = i;
        }
        for(int i=0; i<costs.length;i++){
            int cost = q.get(i).get(0);
            int a = q.get(i).get(1);
            int b = q.get(i).get(2);
            System.out.println(cost+" "+a+" "+b);
            if (find(a, parent) != find(b, parent)){
                System.out.println("ㅇㅇㅇ");
                union(a,b,parent);
                answer += cost;
            }
            
        }
        return answer;
    }
    
}
